### PR TITLE
Plot Step Ahead Projection function

### PR DESCRIPTION
**Pull request by LucieContamin**
- Create a `plot_step_ahead_forecasts()` function (#1)
- Create a vignette illustrating the function

Still to do:

- [ ] The function output Plotly type object, we should add a "interactive" parameter that when set to `TRUE` returns Plotly object, but returns ggplot2 type object when set to `FALSE`.
- [ ] Add Testing

### CODE REVIEW

**Review by harryhoch — COMMENTED** [review]
I think this generally looks good, but I'm concerned that the color brewer plot is unreadable. Can we remove the plot and just say that colors from colorbrewer can be used?

**Comment by LucieContamin** [comment]
> I think this generally looks good, but I'm concerned that the color brewer plot is unreadable. Can we remove the plot and just say that colors from colorbrewer can be used?

Thank you, I can either:
-  remove the plot and change the text to say: `All the palette from RColorBrewer can be used. The possible palette names are accessible by calling: ` and give the command to have all the possible name 
- update the figure parameter to make it easier to read. For example:
image
- or both

**Comment by harryhoch** [comment]
Thanks. Either change would be fine.

File: vignettes/plot_projection.Rmd [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+by setting  the `interactive` parameter to FALSE. See end of the document for examples.
+
+## Load and Filter Data
+
+The package contains 2 datasets that can be used for examples:
+
+- `example_round1.csv`: example of model output for a round associated with the 
+origin date: "2021-03-07" (called "round 1"), target: "incident case", for the
+US national level, from the
+[example-complex-scenario-hub](https://github.com/Infectious-Disease-Modeling-Hubs/example-complex-scenario-hub). The data set also contains an ensemble calculated by applying the
```
Comment (nickreich) [inline vignettes/plot_projection.Rmd:55]: These example files are probably ok as is, but I will mention that there was some discussion at some point about having more centralized example data files, so that we don't have to worry about updating all example data in all these random places when the format updates.

Comment (nickreich) [reply vignettes/plot_projection.Rmd:55]: I don't think we have done anything about this, but tagging @annakrystalli in case there is a better system in place anywhere that we could mimic. Currently the example projections and truth data file are stored as copies in this repo.

File: vignettes/plot_projection.Rmd [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+To plot the model projections for the US, Scenario A:
+
+```{r A_us}
+# Pre-filtering
+projection_data_A_us <- dplyr::filter(projection_data, 
+                                      scenario_id == "A-2021-03-05", 
+                                      location == "US")
+# Class update
+projection_data_A_us <- hubUtils::as_model_out_tbl(projection_data_A_us)
```
Comment (nickreich) [inline vignettes/plot_projection.Rmd:107]: Noting that when I run this line of code, I get an error
```
Error in `[.tbl_df`(tbl, , c(std_colnames[1], names(tbl)[!names(tbl) %in%  : 
  object 'std_colnames' not found
```
That I think has to do with hubUtils needing to be loaded so it can access `std_colnames`. I will add this above with a suggested edit/PR.

Comment (LucieContamin) [reply vignettes/plot_projection.Rmd:107]: I think there is an open issue on hubUtils, [issue 96.](https://github.com/Infectious-Disease-Modeling-Hubs/hubUtils/issues/96). I thought it was solved in an recent version of the package. A solution will be to either update the DESCRIPTION file here to force a latest version of hubUtils or to update the vignette documentation. I can run some tests to identify which version is required

File: R/plot_step_ahead_forecasts.R [hunk 3]
```diff
@@ -0,0 +1,798 @@
...
+  }
+  return(plot_model)
+}
+
+
+#' Basic Plot for model outputs
+#'
+#' Create a simple Plotly time-series plot for model projection outputs.
+#'
+#'@param forecast_data a `model_output_df` object, containing all the required
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/plot_step_ahead_forecasts.R:528]: I agree, it's a good idea. I can change the parameter to "model_output_data" or other, if we prefer something else.

File: R/plot_step_ahead_forecasts.R [hunk 4]
```diff
@@ -0,0 +1,798 @@
...
+#' columns, and a "target_date" and a "model_id" column.
```
Comment (nickreich) [inline R/plot_step_ahead_forecasts.R:529]: Why not leave this more flexible, and allow someone to specify a column name that points to the variable that should be mapped to the x-axis?

File: vignettes/plot_projection.Rmd [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+- `forecast_data`: a `model_output_df` object containing all the Hubverse
+ standard columns, including a `"target_date"`, and a `"model_id"` column in the 
+`hub_mdl_out_df format`. As all forecasts in forecast_data will be plotted, 
+ all filtering needs to happen outside this function.
+- ` truth_data`: a `data.frame` object containing the ground truth data, including 
+the columns: `time_idx` and `value`.
+
+### "Simple" plot
+
+The `"projection_data"` and `"truth_data"` contain information for multiple
```
Comment (nickreich) [inline vignettes/plot_projection.Rmd:96]: Noting that there is inconsistent usage of quoting names of objects inside fixed-width font quotes. e.g.
`"projection_data"` vs `projection_data`.

I've tried to standardize to the convention of only the fixed-width font quotes for everything except for column names, where I include both. Although I'd be happy to follow a more concrete/written standard on this.

File: vignettes/plot_projection.Rmd (hunk 2 above)
Comment (nickreich) [reply vignettes/plot_projection.Rmd:107]: Since it doesn't seem like this has been solved yet, I'm just going to load the `hubUtils` package in the vignette so it builds.

File: R/plot_step_ahead_forecasts.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/plot_step_ahead_forecasts.R:528]: It's ok with me, I was planning to do it later on so thank you very much

File: R/plot_step_ahead_forecasts.R (hunk 4 above)
Comment (LucieContamin) [reply R/plot_step_ahead_forecasts.R:529]: That's a good idea, I can add a `x_col_name` parameter with a default to "target_date".

File: vignettes/plot_projection.Rmd (hunk 5 above)
Comment (LucieContamin) [reply vignettes/plot_projection.Rmd:96]: Sorry for that! I will try to follow that standard in the future.

**Review by nickreich — CHANGES_REQUESTED** [review]
I committed a handful of minor changes to the vignette and one argument name change to the plot_step_ahead_forecasts() function. here are the three items that I see as perhaps things that should be fixed before committing this PR:

- [x] consider a different name for the function, in the spirit of generalizing beyond forecasts. Like `plot_step_ahead_model_output()`?
- [x] per comment on the PR, consider making the "fill_by" argument less restricted to model only.
- [x] add `x_col_name` argument with a default of "target_date".
~~- [ ] possibly consider using more centrally located example files, although I don't think we should let this be a blocker to merging this.~~

**edit**: I am withdrawing my suggestion about centrally located example files.

File: R/plot_step_ahead_forecasts.R [hunk 6]
```diff
@@ -0,0 +1,798 @@
...
+#' @importFrom grDevices col2rgb rgb
+#' @importFrom ggplot2 labs guides
+#'
+#' @export
+#'
+plot_step_ahead_forecasts <- function(
+    forecast_data, truth_data, use_median_as_point = FALSE, plot = TRUE,
+    plot_truth = TRUE, show_legend = TRUE, facet = NULL, facet_scales = "fixed",
+    facet_nrow = NULL, facet_ncol = NULL, facet_title = "top left",
+    interactive = TRUE, fill_by_model = TRUE, pal_color = "Set2",
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/plot_step_ahead_forecasts.R]: I replaced the `fill_by_model` by a `fill_by` parameter taking a valid column name as input and allows users to modify the legend/color setting. By default, the `fill_by` parameter is set to `fill_by = "model_id"`

File: R/plot_step_ahead_forecasts.R (hunk 4 above)
Comment (LucieContamin) [reply R/plot_step_ahead_forecasts.R]: As I am implementing it, I wonder if it makes sense to have the same parameter for the truth_date, with for example a `x_truth_col_name` parameter set by default on "time_idx"

**Comment by LucieContamin** [comment]
> I committed a handful of minor changes to the vignette and one argument name change to the plot_step_ahead_forecasts() function. here are the three items that I see as perhaps things that should be fixed before committing this PR:
> 
>     * [x]  consider a different name for the function, in the spirit of generalizing beyond forecasts. Like `plot_step_ahead_model_output()`?
> 
>     * [x]  per comment on the PR, consider making the "fill_by" argument less restricted to model only.
> 
>     * [x]  add `x_col_name` argument with a default of "target_date".
> 
>     * [ ]  possibly consider using more centrally located example files, although I don't think we should let this be a blocker to merging this.

Thank you for the feedback, I implemented the required changes except for the example files. I am not sure what would be the best solution here.

**Review by nickreich — CHANGES_REQUESTED** [review]
I am basically fine with the content of this PR as is. My final requests are:

- [x] some investigation into why the plotly figure heights are so tall and ideally fixing this so by default they are lower
- [ ] adding at least some unit tests to the code

I do think we should enforce the unit tests piece prior to merging in the current PR, even though the functionality appears to be there. I don't have the time to do this now, but I would be happy at a future time to help brainstorm a list of possible tests that could be put in place here. E.g. one that came to me on this latest re-review is making sure we throw an error if the column specified by x_col_name argument is not a date.

**Comment by saraloo** [comment]
Found this generally very clear and easy to follow. 

I find the attached plot (faceted by model) a bit confusing, with the legend only showing the one model (the first panel). In the interactive plotly, if you then click to not display the model, only the first panel projection disappears (pictured). I would propose to have the default of this plot to not have a legend, but not sure exactly how to set this up practically (tricky with the option default being true for other plots). Another option might be to allow each of the panels to still be a different colour so the legend corresponds to something meaningful?

Screen Shot 2023-10-05 at 12 23 19 PM

File: R/plot_step_ahead_model_output.R [hunk 7]
```diff
@@ -0,0 +1,900 @@
...
+#'
+#' Use Plotly or ggplot2 to plot projection model output with or without
+#' truth data.
+#'
+#' @param all_plot a list with two data frame: one for plain lines,
+#' one for ribbons plotting (in a wide format)
+#' @param all_ens a list with two data frame: one for plain lines,
+#'  one for ribbons plotting (in a wide format) for a unique `model_id` value
+#'  associated with specific color (`ens_color`). NULL is no specific layout
+#'  required
```
Comment (annakrystalli) [inline R/plot_step_ahead_model_output.R:381]: I'm struggling a little to understand the difference between `all_plot` and `all_ens`

File: R/plot_step_ahead_model_output.R [hunk 8]
```diff
@@ -0,0 +1,900 @@
...
+      sharex = TRUE
+    }
+    if (is.null(facet_nrow)) {
+      facet_nrow = 1
+    }
+    subplot <- lapply(facet_value, function(x) {
+      # Data preparation
+      df_point <- all_plot$median[which(all_plot$median[[facet]] == x), ]
+      df_ribbon <- all_plot[names(all_plot) %in% intervals]
+      df_ribbon <- setNames(lapply(df_ribbon, function(df_rib) {
```
Comment (annakrystalli) [inline R/plot_step_ahead_model_output.R:467]:
```suggestion
      df_ribbon <- stats::setNames(lapply(df_ribbon, function(df_rib) {
```

File: R/plot_step_ahead_model_output.R [hunk 9]
```diff
@@ -0,0 +1,900 @@
...
+          x_anchor <- "right"
+        }
+        plot_model <- plotly::layout(
+          plot_model,
+          annotations = list(x = x_title, y = y_title, xref = "paper",
+                             yref = "paper", xanchor = x_anchor,
+                             yanchor = y_anchor, showarrow = FALSE, text = x))
+      }
+      return(plot_model)
+    })
```
Comment (annakrystalli) [inline R/plot_step_ahead_model_output.R:551]: This whole section is quite nested and hard to follow. I wonder if some of it can be modularised and pulled out into smaller more concise functions?

Also although the unnamed function called in `lapply` only has `x` as an argument, there are a lot of objects used from the function environment which also make it a bit confusing and hard to follow.

File: tests/testthat/test-plot_step_ahead_model_output.R [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+
+  plot_test <- plot_step_ahead_model_output(
+    projection_data_A_us, truth_data_us, intervals = 0.9, ens_color = "black",
+    ens_name = "hub-ensemble", facet = "model_id")
+  testthat::expect_equal(
+    strsplit(strsplit(tail(purrr::map(
+      plot_test$x$attrs, "hovertext"), 1)[[1]], "<br>")[[1]][2],
+      "Intervals")[[1]][1], "90%")
+
+  })
```
Comment (annakrystalli) [inline tests/testthat/test-plot_step_ahead_model_output.R:231]: Just to note that snapshots of graphical output can be tested using [`vdiffr`](https://vdiffr.r-lib.org/) package

**Review by annakrystalli — COMMENTED** [review]
Made a number of comments but most are just suggestions. Overall I found nothing critical to resolve but think following some suggestions might make the code easier to follow and reason about.

Also might be a good idea to run `styler::tidyverse_style()` on any `R/*.R` and `testthat/*.R` script to make them more readable.

File: tests/testthat/test-plot_step_ahead_model_output.R [hunk 11]
```diff
@@ -0,0 +1,231 @@
+# Data Preparation
+projection_path <- system.file("example_round1.csv", package = "hubVis")
+projection_data0 <- read.csv(projection_path, stringsAsFactors = FALSE)
+projection_data <- dplyr::mutate(
+  projection_data0, target_date = as.Date(origin_date) + (horizon * 7) - 1)
+projection_data_A_us <- dplyr::filter(projection_data,
+                                      scenario_id == "A-2021-03-05",
+                                      location == "US")
+
+d_proj = rbind(projection_data_A_us,
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply tests/testthat/test-plot_step_ahead_model_output.R]: fixed on [#bbdb005](https://github.com/Infectious-Disease-Modeling-Hubs/hubVis/pull/2/commits/bbdb005b0475c8c21c53f1e2fddd5dcf3a4d8ee5)

File: tests/testthat/test-plot_step_ahead_model_output.R [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+                             model_id = paste0(model_id, "_2")))
+d_proj = hubUtils::as_model_out_tbl(d_proj)
+
+truth_path <- system.file("truth_data.csv", package = "hubVis")
+truth_data <- read.csv(truth_path, stringsAsFactors = FALSE)
+truth_data_us <- dplyr::filter(truth_data, location == "US",
+                               time_idx < min(projection_data$target_date) + 21,
+                               time_idx > "2020-10-01")
+# Test input information
+testthat::test_that("Input parameters", {
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply tests/testthat/test-plot_step_ahead_model_output.R]: I know it's not required but I like to add it because when I am testing it locally, I don't always load the "testthat" package. I removed it to make the script more concise

File: R/plot_step_ahead_model_output.R [hunk 13]
```diff
@@ -0,0 +1,900 @@
+#' Data Process for plotting function
+#'
+#' Data process for plotting function, returns a list with two
+#' data frame: one for plain lines, one for ribbons plotting (in a wide format)
+#'
+#' @param df a `data.frame` object containing the columns: `model_id`,
+#' `output_type_id`, `target_date`, `value`
+#' @param plain_line a `numeric` output_type_id value, value will be used to
+#' create a plain line in the plot. Should be a unique value
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/plot_step_ahead_model_output.R]: I rephrased the documentation

File: R/plot_step_ahead_model_output.R [hunk 14]
```diff
@@ -0,0 +1,900 @@
...
+  plain_df[[x_col_name]] <- as.Date(plain_df[[x_col_name]])
+  # Intervals
+  if (is.null(intervals)) {
+    ribbon_list <- NULL
+  } else {
+    ribbon_list <- lapply(intervals, function(ribbon) {
+      ribbon_df <- df[which(df$output_type_id %in% ribbon), ]
+      ribbon_df <- transform(
+        ribbon_df, output_type_id = ifelse(
+          ribbon_df$output_type_id == min(ribbon), "min", "max"))
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/plot_step_ahead_model_output.R]: Change to `dplyr::near()` and update the dependencies information.

File: R/plot_step_ahead_model_output.R [hunk 15]
```diff
@@ -0,0 +1,900 @@
...
+#'  containing the columns: `time_idx` and `value`.
+#'  Ignored, if `plot_truth = FALSE`
+#' @param plot_truth a `boolean` for showing the truth data in the plot.
+#'  Default to TRUE. Data used in the plot comes from the parameter `truth_data`
+#' @param show_legend a `boolean` for showing the legend in the plot.
+#' @param arguments list of others Plotly parameters.
+#' @param x_col_name column name containing the date information for the x-axis.
+#' By default, "time_idx".
+#'
+#' @importFrom plotly add_trace layout
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/plot_step_ahead_model_output.R]: I know it's not necessary to add the importFrom but I use it to track my package import more easily. I am ok to remove it if necessary.

File: tests/testthat/test-plot_step_ahead_model_output.R (hunk 10 above)
Comment (LucieContamin) [reply tests/testthat/test-plot_step_ahead_model_output.R:231]: Thanks I was looking into it and thinking to implement it in the next version

File: R/plot_step_ahead_model_output.R (hunk 9 above)
Comment (LucieContamin) [reply R/plot_step_ahead_model_output.R]: I agree, it will take me some time to test and implement. I am happy to think about it for a future version.

File: R/plot_step_ahead_model_output.R [hunk 16]
```diff
@@ -0,0 +1,900 @@
...
+#' @importFrom cli cli_abort cli_warn
+#' @importFrom scales percent
+#' @importFrom methods show
+#' @importFrom stats setNames
+#' @importFrom RColorBrewer brewer.pal
+#' @importFrom grDevices col2rgb rgb colors
+#' @importFrom ggplot2 labs guides
+#'
+#' @export
+#'
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/plot_step_ahead_model_output.R:77]: I add a very small example, I am happy to add more examples later on. I also add a comment that multiple examples are available in the vignette. 
Also, I split the file into 2 as asked.
